### PR TITLE
Expand crypt module to cover Notices, Actions & Topics

### PR DESCRIPTION
**Pull request by anthonyryan1**
Improves compatibility with mircryption.

Notes:
1. I know the use of `OnRaw` is not suggested, but I cannot find a more appropriate way to intercept the incoming topic message.
2. There's some duplication here, but when I added functions for the individual pieces of functionality needed to handle outgoing messages, it did not seem to significantly improve readability.

I'm open to all comments and suggestions.

### CODE REVIEW

**Comment by DarthGandalf** [comment]
`OnTopic`?

File: modules/crypt.cpp [hunk 1]
```diff
@@ -84,18 +87,144 @@ class CCryptMod : public CModule {
...
+			return HALTCORE;
+		}
+
+		return CONTINUE;
+	}
+
+	virtual EModRet OnUserAction(CString& sTarget, CString& sMessage) override {
+		sTarget.TrimLeft(NickPrefix());
+
+		if (sMessage.StartsWith("``")) {
```
Comment (DarthGandalf) [inline modules/crypt.cpp]: if (s.TrimPrefix(..))

**Comment by DarthGandalf** [comment]
Looks like you're not open to comments :)

**Comment by anthonyryan1** [comment]
With all due respect, your comment in this issue wasn't exactly meaningful. What about OnTopic?

The other part, was admittedly my fault. I didn't recall seeing your comment about switching the `sTarget.TrimLeft(NickPrefix())` with `sTarget.TrimPrefix()`. If you can elaborate on what you mean I can ensure both of your requests are incorporated.

**Comment by DarthGandalf** [comment]
OnTopic was response for "but I cannot find a more appropriate way to intercept the incoming topic message."

**Comment by anthonyryan1** [comment]
At least when I tested, the raw 332 was not corrected by OnTopic alone. I can chekc if that has changed since my prior testing. `OnTopic` is  used, but without the `OnRaw` it wasn't working fully.

Perhaps a better solution to that would be to ensure that OnTopic does intercept the 332s as well as what it does currently.

**Comment by DarthGandalf** [comment]
Oh, I didn't notice `sTarget.TrimLeft` before, thanks for fixing it. `TrimLeft` does something completely different. Speaking about `sTarget.TrimLeft`, there is a similar one added below in this PR.
My comment was about `sMessage.StartsWith("``")`.

Regarding `OnTopic`, I see... OnTopic handles cases when it is changed, 332 is for case if client requests `/topic` and joins channel.

May you rebase/squash commits?

**Comment by anthonyryan1** [comment]
I hope to get this rebased tomorrow, sorry about the delay.

**Comment by anthonyryan1** [comment]
Rebased & tested.

Let me know if you have any comments of concerns.

File: modules/crypt.cpp [hunk 2]
```diff
@@ -84,22 +86,144 @@ class CCryptMod : public CModule {
 		return CONTINUE;
 	}
 
+	EModRet OnUserNotice(CString& sTarget, CString& sMessage) override {
+		sTarget.TrimLeft(NickPrefix());
```
Comment (DarthGandalf) [inline modules/crypt.cpp]: TrimLeft or TrimPrefix?

**Comment by anthonyryan1** [comment]
It should have indeed been TrimPrefix. Somehow that change regressed while rebasing.

**Pull request merged**